README.md

```diff
@@ -1,21 +1,64 @@
 [![Build Status](http://ci.bazel.io/buildStatus/icon?job=BUILD_file_generator)](http://ci.bazel.io/job/BUILD_file_generator)
 
-# BUILD File Generator
+# BUILD File Generator (BFG)
 
-BUILD File Generator generates Bazel BUILD files for Java code.
+BUILD File Generator, or BFG for short, is a tool for generating Bazel BUILD files from one's source code. Instead of manually going through your project and creating BUILD files and their appropriate build rules, you can rely on BFG to do it for you! 
 
-1. It reads all `.java` files, and extracts the class dependency graph.
-2. Computes the strongly connected components of the graph.
-3. For each component, creates a `java_library` rule.
+An example project can be found here. TODO
+
+### Why is it useful?
+
+Bazel promises fast and correct builds, _especially_ for incremental builds. Rather than rebuild your entire codebase, Bazel _only_ rebuilds what is necessary, the targets that you have changed. For the remainder of your code, it relies on cached versions. 
+
+There is one caveat to this fast incremental performance, though. It is dependent on the granularity of your targets. At one extreme, the least granular end, you can define a single glob for your entire project. This would be the quickest way to get your project using Bazel. However, it comes at the cost of eradicating any of Bazel's incremental build performance. On the other end, you can try to manually define the most granular targets possible, a single rule per file, collapsing any cyclic dependencies into the same target. This would ensure you make use of Bazel's blazingly fast incremental builds. However, it is extremely tedious, and somewhat annoying to maintain.
+
+BFG automates this difficult tango.
 
 ## Usage
 
-TODO
+BFG is composed of two general components
+
+1. Language specific parsers (LSP)
+2. The BFG binary
+
+The LSPs read your source code and generate a class dependency graph in the form of a protobuf. To generate your BUILD files, you pass the generated protobuf into the BFG binary.
+
+### Step 0: Installation Instructions
+
+To install BFG...TODO
+
+### Step 1: Using LSPs to generate dependency graphs
+
+TODO(bazel-devel): add explanation and valid example arguments.
+
+```bash
+bazel run //lang:JavaSourceFileParserCli <TODO>
```

Comment:
@cgrushko how do you actually run the `JavaSourceFileParserCli`? For example what is the command line args for for `google-java-format`?

I wasn't able to immediately figure out what the command line format was, or how the input proto should be formatted.

Reply:
For google-java-format it should probably be
    
    bazel run //lang/java/src/main/java/com/google/devtools/build/bfg:JavaSourceFileParserCli --roots=core/src/main/java,core/src/test/java $(find core/src/main/java/ core/src/test/java/ -name \*.java) > bfg.bin

It's a bit silly that we have to pass both roots and files, but it comes from Google where you don't want to read the entire java tree just because it's the root. We can perhaps improve this by scanning all of roots if the user doesn't provide source files, but it's for another day.

It should print usage information if run without flags (but I haven't tried it lately). Does it now?
By format, what do you mean?